File: v8-core/src/audit/reconciliation.rs
//! Reconciliation Receipt & Clone Collapse Auditor (D-132, Rule 20/22).

use crate::opportunity::reconcile::{ReconciliationReceipt, ReconciledOpportunityState};

pub struct ReconciliationAuditor;

impl ReconciliationAuditor {
    /// Validates receipt cryptographic authenticity and clone collapse proof integrity.
    pub fn audit_receipt(
        state: &ReconciledOpportunityState,
        receipt: &ReconciliationReceipt,
    ) -> Result<(), String> {
        if state.receipt_id != receipt.receipt_id {
            return Err("RECEIPT_ID_MISMATCH".to_string());
        }

        if receipt.compute_id() != receipt.receipt_id {
            return Err("RECEIPT_PAYLOAD_TAMPERED".to_string());
        }

        if state.compute_id() != state.reconciled_id {
            return Err("STATE_IDENTITY_TAMPERED".to_string());
        }

        // Validate effective observer count matches collapse proofs
        let proof_effective_sum: f64 = receipt
            .collapse_proofs
            .iter()
            .map(|p| p.normalized_effective_weight)
            .sum();

        if (receipt.effective_observer_count - proof_effective_sum).abs() > 1e-6 {
            return Err(format!(
                "CLONE_COLLAPSE_INTEGRITY_FAILURE: EffectiveCount={}, ProofSum={}",
                receipt.effective_observer_count, proof_effective_sum
            ));
        }

        Ok(())
    }
}

File: v8-next/src/v8_next/audit/reconciliation.py
"""Reconciliation receipt and clone-collapse auditor — port of v8-core/src/audit/reconciliation.rs.

Faithful in structure: the receipt id must match the state, the receipt payload must hash back
to its own id, the state identity must hash back to itself, and the effective observer count
must equal the sum of the collapse-proof weights (no clone survives uncollapsed).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

CLONE_COLLAPSE_EPSILON = 1e-6


class CollapseProof(Protocol):
    normalized_effective_weight: float


@dataclass(frozen=True)
class ReconciliationReceipt:
    receipt_id: str
    compute_id: str
    effective_observer_count: float
    collapse_proofs: tuple[CollapseProof, ...]


@dataclass(frozen=True)
class ReconciledOpportunityState:
    receipt_id: str
    compute_id: str
    reconciled_id: str


class ReconciliationFailure(ValueError):
    """Raised when a receipt or a clone-collapse proof does not verify (Rules 20/22)."""


class ReconciliationAuditor:
    """Port of ``v8_core::audit::ReconciliationAuditor``."""

    @staticmethod
    def audit_receipt(
        state: ReconciledOpportunityState,
        receipt: ReconciliationReceipt,
    ) -> None:
        if state.receipt_id != receipt.receipt_id:
            raise ReconciliationFailure("RECEIPT_ID_MISMATCH")
        if receipt.compute_id != receipt.receipt_id:
            raise ReconciliationFailure("RECEIPT_PAYLOAD_TAMPERED")
        if state.compute_id != state.reconciled_id:
            raise ReconciliationFailure("STATE_IDENTITY_TAMPERED")
        proof_sum = float(sum(p.normalized_effective_weight for p in receipt.collapse_proofs))
        if abs(receipt.effective_observer_count - proof_sum) > CLONE_COLLAPSE_EPSILON:
            raise ReconciliationFailure(
                "CLONE_COLLAPSE_INTEGRITY_FAILURE: "
                f"EffectiveCount={receipt.effective_observer_count}, ProofSum={proof_sum}"
            )
